### PR TITLE
output/eve: add 'verdict' field to 'alert' and 'drop' events -  v8

### PR DESCRIPTION
**Pull request by jufajardini**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5464

Previous PR: https://github.com/OISF/suricata/pull/8596

Sharing this now to clarify one aspect, and check if this is going on the right direction.

Describe changes from previous PR:
- add verdict as a new output event
- use flags to set configuration options for `drop` event (haven't touched the `flows all/start` one, as this was leading to failures, but can make it a TODO
- drop idea of having `alert.action` as optional
- incorporate verdict format suggestions from previous PR

TODOs:

- add more examples in the documentation
- document the json schema
- document verdict configs for suricata.yaml (userguide)

### Provide values to any of the below to override the defaults.

```
SV_BRANCH=pr/1293
```
https://github.com/OISF/suricata-verify/pull/1293

Outputs examples:
```json
# ids mode, reject, verdict event:
{
  "timestamp": "2004-05-13T10:17:08.993643+0000",
  "flow_id": 1055224944964893,
  "pcap_cnt": 6,
  "event_type": "verdict",
  "src_ip": "65.208.228.223",
  "src_port": 80,
  "dest_ip": "145.254.160.237",
  "dest_port": 3372,
  "proto": "TCP",
  "pkt_src": "wire/pcap",
  "verdict": {
    "action": "alert",
    "reject-target": "source",
    "reject": "[tcp-reset, icmp-prohib, user defined]"
  }
}

# ids mode, reject, alert event (non-interesting portions removed):
{
  "timestamp": "2004-05-13T10:17:08.993643+0000",
  "flow_id": 1055224944964893,
  "pcap_cnt": 6,
  "event_type": "alert",
  "src_ip": "65.208.228.223",
  "src_port": 80,
  "dest_ip": "145.254.160.237",
  "dest_port": 3372,
  "proto": "TCP",
  "pkt_src": "wire/pcap",
  "alert": {
    "action": "blocked",
    "gid": 1,
    "signature_id": 1000001,
    "rev": 0,
    "signature": "",
    "category": "",
    "severity": 3
  },
...
  "verdict": {
    "action": "alert",
    "reject-target": "source",
    "reject": "[tcp-reset, icmp-prohib, user defined]"
  }
}
# ids mode, alert:
{
  "timestamp": "2004-05-13T10:17:08.783340+0000",
  "flow_id": 1055224944964893,
  "pcap_cnt": 5,
  "event_type": "verdict",
  "src_ip": "65.208.228.223",
  "src_port": 80,
  "dest_ip": "145.254.160.237",
  "dest_port": 3372,
  "proto": "TCP",
  "pkt_src": "wire/pcap",
  "verdict": {
    "action": "alert"
  }
}
# ips mode, drop event:
{
  "timestamp": "2016-07-13T22:42:07.388277+0000",
  "flow_id": 2019293786341530,
  "pcap_cnt": 8,
  "event_type": "drop",
  "src_ip": "10.16.1.11",
  "src_port": 54186,
  "dest_ip": "82.165.177.154",
  "dest_port": 80,
  "proto": "TCP",
  "pkt_src": "wire/pcap",
  "direction": "to_server",
  "drop": {
    "len": 40,
    "tos": 0,
    "ttl": 64,
    "ipid": 36762,
    "tcpseq": 2059970575,
    "tcpack": 2548486954,
    "tcpwin": 237,
    "syn": false,
    "ack": true,
    "psh": false,
    "rst": false,
    "urg": false,
    "fin": true,
    "tcpres": 0,
    "tcpurgp": 0,
    "reason": "flow drop"
  },
  "verdict": {
    "action": "drop"
  }
}
# ips mode, alert event, drop rule:
{
  "timestamp": "2016-07-13T22:42:07.199844+0000",
  "flow_id": 2019293786341530,
  "pcap_cnt": 4,
  "event_type": "alert",
  "src_ip": "10.16.1.11",
  "src_port": 54186,
  "dest_ip": "82.165.177.154",
  "dest_port": 80,
  "proto": "TCP",
  "pkt_src": "wire/pcap",
  "alert": {
    "action": "blocked",
    "gid": 1,
    "signature_id": 2,
    "rev": 1,
    "signature": "",
    "category": "",
    "severity": 3
  },
...
  "verdict": {
    "action": "drop"
  }
}
```

### CODE REVIEW

File: src/output-json-drop.c [hunk 1]
```diff
@@ -60,7 +61,8 @@
 
 #define MODULE_NAME "JsonDropLog"
 
-#define LOG_DROP_ALERTS 1
+#define LOG_DROP_ALERTS  BIT_U8(1)
+#define LOG_DROP_VERDICT BIT_U8(2)
```
Comment (jufajardini) [inline src/output-json-drop.c:65]: Changed this, tried adding the flows start/all option here, but led to failures. Can make this a TODO.

File: src/output-json-verdict.c [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            JB_SET_STRING(jb, "action", "alert");
+        }
+        if (p->action & ACTION_REJECT || p->action) {
+            JB_SET_STRING(jb, "reject-target", "source");
+        } else if (p->action & ACTION_REJECT_DST) {
+            JB_SET_STRING(jb, "reject-target", "destination");
+        } else if (p->action & ACTION_REJECT_BOTH) {
+            JB_SET_STRING(jb, "reject-target", "both");
+        }
+        JB_SET_STRING(jb, "reject", "[tcp-reset, icmp-prohib, user defined]");
```
Comment (jufajardini) [inline src/output-json-verdict.c:72]: This is the bit I wasn't sure about the approach, @victorjulien .  The examples from https://github.com/OISF/suricata/pull/8596#discussion_r1197787688 show all options. If that's the case, would this work? Or do we want to check what is actually being done by Suri, and output only that option? If so, I'm done sure how to figure out when will it be `user defined`... 

The other two I figured I could decide by checking the packet proto.

File: doc/userguide/output/eve/eve-json-format.rst [hunk 3]
```diff
@@ -147,13 +141,76 @@ the signature.
...
+::
+
+  "action":"allowed"
+
+Action is set to "allowed" unless a rule used the "drop" action and Suricata is
+in IPS mode, or when the rule used the "reject" action. It is important to note
+that this does not necessarily indicate the final verdict for a given packet or
+flow, since one packet may match on several rules.
+
+Verdict Field
```
Comment (victorjulien) [inline doc/userguide/output/eve/eve-json-format.rst:160]: this is not a field, but an object containing other fields, right?

File: doc/userguide/output/eve/eve-json-format.rst [hunk 4]
```diff
@@ -147,13 +141,76 @@ the signature.
...
+
+Examples
+~~~~~~~~
+
+::
+
+   "verdict": {
+      "action": "alert"
+      "reject-target": "both"
+      "reject": [ "tcp-reset", "icmp-prohib", "user defined" ]
```
Comment (victorjulien) [inline doc/userguide/output/eve/eve-json-format.rst:199]: the value here should be a json array of strings

File: src/output-json-verdict.c (hunk 2 above)
Comment (victorjulien) [reply src/output-json-verdict.c:72]: this shouldn't be a single string, but a json array of strings

Comment (victorjulien) [reply src/output-json-verdict.c:72]: "user defined" can be ignored, it was a place holder for some future ideas

File: src/output-json-verdict.c [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+ * \param p            Pointer to the packet that stores the info on verdict
+ *
+ * \return 0 on success
+ */
+static int JsonVerdictLogger(ThreadVars *tv, void *thread_data, const Packet *p)
+{
+    JsonVerdictLogThread *vlt = thread_data;
+
+    int r = VerdictJson(vlt, p);
+
```
Comment (victorjulien) [inline src/output-json-verdict.c:120]: remove newline

File: src/output-json-verdict.c [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    if (r < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int JsonVerdictLogCondition(ThreadVars *tv, void *thread_data, const Packet *p)
+{
+    if (p->alerts.cnt > 0) {
```
Comment (victorjulien) [inline src/output-json-verdict.c:130]: so verdict records are only written for alerts... this excludes drops from outside the detect engine (e.g. exception policy), also "pass" would be excluded, right?

**Review by victorjulien — CHANGES_REQUESTED** [review]
see inline comments

File: src/output-json-verdict.c (hunk 6 above)
Comment (jufajardini) [reply src/output-json-verdict.c:130]: Right on cue for both aspects, I hadn't considered them >__<'

File: doc/userguide/output/eve/eve-json-format.rst (hunk 3 above)
Comment (jufajardini) [reply doc/userguide/output/eve/eve-json-format.rst:160]: Right, I was seeing it as a field because it's part of 'alert', but... that's indeed wrong.

Comment (jufajardini) [reply doc/userguide/output/eve/eve-json-format.rst:160]: also: update description, there are more fields in verdict now.

File: src/output-json-verdict.c (hunk 2 above)
Comment (jufajardini) [reply src/output-json-verdict.c:72]: I see, thanks.

**Comment by jufajardini** [comment]
Replaced by: https://github.com/OISF/suricata/pull/9216

**Pull request closed without merging**